pyquil/pyqvm.py

```diff
@@ -0,0 +1,459 @@
+            self.program_counter += 1
+
+        elif isinstance(instruction, Jump):
+            # unconditional Jump; go directly to Label
+            self.program_counter = self.find_label(instruction.target)
+
+        elif isinstance(instruction, JumpTarget):
+            # Label; pass straight over
+            self.program_counter += 1
+
+        elif isinstance(instruction, JumpConditional):
+            # JumpConditional; check classical reg
+            x = instruction.condition  # type: MemoryReference
+            cond = self.ram[x.name][x.offset]
+            if not isinstance(cond, (bool, np.bool, np.int8)):
+                raise ValueError("{} requires a data type of BIT; not {}"
+                                 .format(instruction.op, type(cond)))
+            dest_index = self.find_label(instruction.target)
+            if isinstance(instruction, JumpWhen):
+                jump_if_cond = True
+            elif isinstance(instruction, JumpUnless):
+                jump_if_cond = False
+            else:
+                raise TypeError("Invalid JumpConditional")
+
+            if not (cond ^ jump_if_cond):
+                # jumping: set prog counter to JumpTarget
+                self.program_counter = dest_index
+            else:
+                # not jumping: hop over this JumpConditional
+                self.program_counter += 1
+
+        elif isinstance(instruction, UnaryClassicalInstruction):
+            # UnaryClassicalInstruction; set classical reg
+            target = instruction.target  # type:MemoryReference
+            old = self.ram[target.name][target.offset]
+            if isinstance(instruction, ClassicalNeg):
+                if not isinstance(old, (int, float, np.int, np.float)):
+                    raise ValueError("NEG requires a data type of REAL or INTEGER; not {}"
+                                     .format(type(old)))
+                self.ram[target.name][target.offset] *= -1
+            elif isinstance(instruction, ClassicalNot):
+                if not isinstance(old, (bool, np.bool)):
+                    raise ValueError("NOT requires a data type of BIT; not {}"
+                                     .format(type(old)))
+                self.ram[target.name][target.offset] = not old
+            else:
+                raise TypeError("Invalid UnaryClassicalInstruction")
+
+            self.program_counter += 1
+
+        elif isinstance(instruction, (LogicalBinaryOp, ArithmeticBinaryOp, ClassicalMove)):
+            left_ind = instruction.left  # type: MemoryReference
+            left_val = self.ram[left_ind.name][left_ind.offset]
+            if isinstance(instruction.right, MemoryReference):
+                right_ind = instruction.right  # type: MemoryReference
+                right_val = self.ram[right_ind.name][right_ind.offset]
+            else:
+                right_val = instruction.right
+
+            if isinstance(instruction, ClassicalAnd):
```

Comment:
This is an approximation of the behavior that should happen. (e.g., it should not be possible to add two bits) Each instruction has roughly an "addressing mode" of supported data types.